plugin/mlsdbonlinelocator.cpp

```diff
@@ -88,8 +111,89 @@ void MlsdbOnlineLocator::defaultVoiceModemChanged(const QString &modem)
     setupSimManager();
 }
 
-bool MlsdbOnlineLocator::findLocation(const QList<MlsdbProvider::CellPositioningData> &cells)
+QPair<QDateTime, QVariantMap> MlsdbOnlineLocator::buildLocationQuery(
+        const QList<MlsdbProvider::CellPositioningData> &cells,
+        const QPair<QDateTime, QVariantMap> &oldQuery) const
+{
+    static bool waitForWlanInfo = true;
+    const QDateTime currDt = QDateTime::currentDateTimeUtc();
+    QVariantMap map;
+    map.unite(cellTowerFields(cells));
+    map.unite(wlanAccessPointFields());
+
+    if (map.isEmpty()) {
+        // no field data(cell, wifi) available
+        qCDebug(lcGeoclueMlsdbOnline) << "No field data(cell, wifi) available for MLS online request";
+    } else if (!map.contains(QStringLiteral("wifiAccessPoints")) && waitForWlanInfo) {
+        // it can take some time to receive wlan network info.
+        // the MLS online lookup is far more accurate if we have some wlan network info to provide.
+        // so, if we have no wlan info, and this was the first request, don't do an online request yet.
+        qCDebug(lcGeoclueMlsdbOnline) << "No wifi data available for MLS online request, postponing";
+        waitForWlanInfo = false;
+    } else {
+        map.unite(globalFields());
+        map.unite(fallbackFields());
+
+        // Only send the query if we have more information than previously
+        // or if sufficient time has passed since the last query we performed.
+        const bool firstTimeQuery = oldQuery.first.isNull() || oldQuery.second.isEmpty();
+        const bool intervalExceeded = oldQuery.first.isNull() || oldQuery.first.msecsTo(currDt) >= m_adaptiveInterval;
+        const bool moreInfo = map.keys().size() > oldQuery.second.keys().size();
+        const bool newCells = cellIdsFromQueryData(oldQuery.second) != cellIdsFromQueryData(map);
+
+        if (firstTimeQuery || intervalExceeded || moreInfo || newCells) {
+            // adaptively back-off future requests to avoid server-side throttling.
+            static quint32 backOffFactor = 8;
```

Comment:
Doesn't this make the backOffFactor to be always 8 when we do any further calculations later?

Reply:
function static variable so it will retain its value between function invocations

Reply:
Hah, I'm blind ;)